### DISCUSSION TITLE
Active Router Health endpoint

### DISCUSSION THREAD

**Discussion by sanchaymittal** [connext/nxtp#738] [Specs]
Problem: 
- Need transfer size based on active routers.

Example case/Test case:
- Nomad or xpolli can use this endpoint to figure out max transfer size based on active routers.

Owner:
@sanchaymittal @base0010 

Proposal:

T0 [sdk]
- `getActiveRouterHealth()`: Call getRouterStatus() get live routers and call subgraphs to getRouterLiq().
- `getHealthFromEndpoint(): Call endpoint to get router health with available liq.

T1 [sdk-server]
- `getRouterHealth()`: Calls `getActiveRouterHealth()` to get router health and liq.
- expose this endpoint.

T0 [EC2][Cloudflare]
- host the server on ec2
- connect ec2 with Cloudflare

Test Cases:
```
- should work with endpoint
- should work if no endpoint
- should work if endpoint errors
```

**Comment by base0010** [comment]
[CF]:
* what kind of auth/endpoint does this need on the ec2 side? or none ? ie .is the ec2 internal only and exposed by CF to handle loadbalancing et al. 
* CF endpoints ie. api spec:
 -    / 
       - returns all router liveliness & liquidity indexed on address
 -    /?routerAddress=0x...abcdef 
       - returns only that router's liveliness & liquidity

**Reply by sanchaymittal** [reply]
- Doesn't really need any auth for the ec2 endpoint, it's fine for anyone to access it.

**Comment by LayneHaber** [comment]
Is the reason we aren't using the subgraph for this because of the liquidity caching, the latency, or both? What would the body of the endpoint look like?

**Reply by rhlsthrm** [reply]
Nomad and Coinhippo want an API to call to get available liquidity on a path. It's pretty intensive to regenerate this from subgraph queries every time, so it would be easier to cache this data on a backend. This also builds towards the explorer backend stuff.

**Reply by LayneHaber** [reply]
Is this separate from the router itself then? Without having an idea of the reserved/already bid liquidity this won't return accurate results

**Reply by rhlsthrm** [reply]
At least the "idle" liquidity will give us a better idea than right now, but we can also ask the router to tell us this info and it can tell us what the available is.

**Reply by rhlsthrm** [reply]
The main use case is to get this info for Nomad, so if you have better ideas let's discuss @LayneHaber !

**Reply by LayneHaber** [reply]
Right, I think it depends on how reliable we want the data to be. The most reliable would be to have the routers self-report this way they would include any information about reserved liquidity. The problem then is how do you deal with misreported information, and do you then need to set up a messaging system where routers publish to some NATS subject that's then polled by the service

**Reply by sanchaymittal** [reply]
we can query it from the subgraph directly, don't need it from the router. 
We just need to check if the router is online or not, and then get the liquidity for those routers.

**Reply by rhlsthrm** [reply]
Or the router can report it, then the SDK can validate it before sending.

**Reply by sanchaymittal** [reply]
Yupp that too

**Reply by base0010** [reply]
** So I think we decided to just use AWS native loadbalancing etc. since this already will exist as an EC2.

 ** I do agree that there is mitigable concern with the router self reporting it's liquidity. Mitigation is to double check subgraph or just check subgraph instead. 
 
 ** It would be ideal if you could have them sign this response in a way that convicts/slashed them for reporting incorrect liquidity. But that very much feels like a v2 idea.

**Reply by LayneHaber** [reply]
Staking/slashing for misreporting liquidity is probably not enforceable since its all offchain, I think if we need this to be for other things outside of the sdk (and it will grow to become the subgraph collation layer), then we should probably just use the subgraph but acknowledge it may not be in line with what the routers are saying.

I guess whether or not that's acceptable really depends on the other use cases (i.e. what is Nomad planning on using this for? For coinhippo this is likely fine.)

**Reply by jakekidd** [reply]
> Staking/slashing for misreporting liquidity is probably not enforceable since its all offchain

It is if they sign a warrant when reporting it. Warrant being something that (we know) can be used to slash if the thing turns out to not be true

But also personally, I don't feel it's necessary. Routers can't bid if they don't have the money, and routers would want to report that they have as much as they do. Don't really see the incentive to lie?

EDIT: Actually maybe there's an incentive to lie and say they have a ton more just to get more bids (esp. cause multipath as well now).

**Reply by jakekidd** [reply]
We could aggregate the router API response and compare it actively to cached subgraph queries on the backend for our endpoint?
